sdk/iot/hub/tests/cmocka/test_az_iot_hub_client.c

```diff
@@ -741,6 +761,107 @@ static void test_az_iot_hub_client_properties_find_value_match_end_fail(void** s
+      (size_t)az_span_length(test_value_two));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_OK);
+  assert_memory_equal(
+      az_span_ptr(pair_out.key), az_span_ptr(test_key_three), (size_t)az_span_length(test_key_three));
+  assert_memory_equal(
+      az_span_ptr(pair_out.value),
+      az_span_ptr(test_value_three),
+      (size_t)az_span_length(test_value_three));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_ERROR_EOF);
+}
+
+static void test_az_iot_hub_client_properties_next_twice_succeed(void** state)
+{
+  (void)state;
+
+  az_span test_span = az_span_from_str(TEST_KEY_VALUE_TWO);
+  az_iot_hub_client_properties props;
+  assert_int_equal(az_iot_hub_client_properties_init(&props, test_span), AZ_OK);
+
+  az_pair pair_out;
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_OK);
+  assert_memory_equal(
+      az_span_ptr(pair_out.key), az_span_ptr(test_key_one), (size_t)az_span_length(test_key_one));
+  assert_memory_equal(
+      az_span_ptr(pair_out.value),
+      az_span_ptr(test_value_one),
+      (size_t)az_span_length(test_value_one));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_OK);
+  assert_memory_equal(
+      az_span_ptr(pair_out.key), az_span_ptr(test_key_two), (size_t)az_span_length(test_key_two));
+  assert_memory_equal(
+      az_span_ptr(pair_out.value),
+      az_span_ptr(test_value_two),
+      (size_t)az_span_length(test_value_two));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_ERROR_EOF);
+
+  //Reset to beginning of span
+  assert_int_equal(az_iot_hub_client_properties_init(&props, test_span), AZ_OK);
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_OK);
+  assert_memory_equal(
+      az_span_ptr(pair_out.key), az_span_ptr(test_key_one), (size_t)az_span_length(test_key_one));
+  assert_memory_equal(
+      az_span_ptr(pair_out.value),
+      az_span_ptr(test_value_one),
+      (size_t)az_span_length(test_value_one));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_OK);
+  assert_memory_equal(
+      az_span_ptr(pair_out.key), az_span_ptr(test_key_two), (size_t)az_span_length(test_key_two));
+  assert_memory_equal(
+      az_span_ptr(pair_out.value),
+      az_span_ptr(test_value_two),
+      (size_t)az_span_length(test_value_two));
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_ERROR_EOF);
+}
+
+static void test_az_iot_hub_client_properties_next_empty_succeed(void** state)
+{
+  (void)state;
+
+  uint8_t empty_prop_buf[8] = { 0 };
+  az_span test_span = az_span_init(empty_prop_buf, 0, _az_COUNTOF(empty_prop_buf));
+  az_iot_hub_client_properties props;
+  assert_int_equal(az_iot_hub_client_properties_init(&props, test_span), AZ_OK);
+
+  az_pair pair_out;
+
+  assert_int_equal(az_iot_hub_client_properties_next(&props, &pair_out), AZ_ERROR_EOF);
+}
+
```

Comment:
Should there be a test for when the properties are incorrectly formatted?
Would we want to have that validation in the function? I know it would be expensive.

Reply:
I went back and forth on this. But for now, I don't expect hub to be giving us malformed properties. If it's user properties that the user has allocated in the device code (not input from hub), then I believe the burden is on the developer to have them properly formatted. In order to be slim and fast, I think here the developer should accept some responsibility.